Fix #19935: Part of old multiplayer name remains when entering a shorter name

Chase-Percy opened the pull request:
Fixes #19935

Proposed Solution:

- Memset _playerName with null terminators before the text string view is copied into it.

Note:
-  I'm unsure if it would update the network version since it shouldn't change anything.

CookiePLMonster commented:
Unless your PR is WIP, there seems to be no file overlap.

ZehMatt commented:
> This is a good opportunity to turn `_playerName` into a `u8string` instead IMO - not only it lifts an artificial limit while fixing the bug you are addressing, but should also make the code clearer. It's a `.cpp` only UI class so I doubt you need to worry about the network version.

I agree, we should use u8string here and avoid having to use C style APIs in general.

Chase-Percy commented:
Ah okay yeah a u8string would be clearer. I wasn't sure if there was a reason for the limit on the array. Is there any external documentation somewhere like a confluence or anything? Kinda hard when new to the codebase to know why things were done in a certain way like limitations etc.

Chase-Percy commented:
Ok cool, good to know going forward! I'll rectify tomorrow

Chase-Percy commented:
I have updated it to a u8string.

Here are some things worth mentioning:
- I set a value for the length of the player name. I based it on the old truncated value at 32 because otherwise, the names don't fit nicely in the in-game UI.
- I used the fill constructor to set the string size so the underlying c_str isn't invalidated. I was having issues if you opened the window with a short name and then expanded it to be the max size or enough to trigger the underlying c_str to be reallocated.
- I made the WindowStartTextbox function take a const char* so a c_str can be passed in without const casting it.
- I removed the old empty player name check since the text.empty() guard always procs before it can get there.

duncanspumpkin commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -267,22 +268,17 @@ class ServerListWindow final : public Window
-                    gConfigNetwork.PlayerName = _playerName;
-                    ConfigSaveDefault();
-                }
+                _playerName = temp;
```
This will leak the old pointer. You must reset the widget pointer after this

Chase-Percy replied:
Yea, I had thought I solved this by setting the underlying c_str to be the same size as the max name length so on it would copy and not proc a realloc, but I thought about it, and this would be vendor implementation specific and is not defined in the standard. I think the approach you suggest is the safest: [3d02334](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/3d02334f2dc3124cf3cc737ee50f49aaa3a849de).

Broxzier commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -267,22 +268,17 @@ class ServerListWindow final : public Window
             return;
 
-        std::string temp = static_cast<std::string>(text);
+        u8string temp = static_cast<u8string>(text);
```
The cast already states the type explicitly, so no need to state this twice.

```suggestion
        auto temp = static_cast<u8string>(text);
```

Chase-Percy replied:
[6f4df6e](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/6f4df6e38ba05451cee9494f4e70063d5d5a10cc)

Broxzier commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -267,22 +268,17 @@ class ServerListWindow final : public Window
-                }
+                _playerName = temp;
+                gConfigNetwork.PlayerName = _playerName;
+                ConfigSaveDefault();
```
More of a remark than something to change right now. This gets called with every character you type. I think it would be nicer to only save it when the input box loses focus, or when Enter is pressed or something.

Chase-Percy replied:
I agree, I gave it a shot here so it saves on window close or loss of focus: [c6e2546](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/c6e2546accada9b558422bde983a43943cdd5d84)

Broxzier replied:
Nice to see you picked this up. I left a few notes on the implementation.

Chase-Percy commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -33,6 +33,8 @@
 #    define WHEIGHT_MAX 800
 #    define ITEM_HEIGHT (3 + 9 + 3)
 
+constexpr size_t PLAYER_NAME_LENGTH = 32;
```
Yea sorry, still getting used to the conventions here: [43c8440](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/43c84404f6e04ff5cf40990ad6cce15503633d3e)

CookiePLMonster commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -76,21 +78,32 @@ void JoinServer(std::string address);
 {
 private:
-    char _playerName[32 + 1] = {};
+    u8string _playerName = {};
```
```suggestion
    u8string _playerName;
```

Chase-Percy replied:
Yup, good point [38083b2](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/38083b2cd74b2329d29f68d781a86eb8e0a86d83)

CookiePLMonster reviewed (commented):
Left just one nitpick, looks good to me otherwise.

Broxzier commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -76,21 +78,32 @@ void JoinServer(std::string address);
     bool _showNetworkVersionTooltip = false;
     std::string _version;
 
+    void CheckTextBoxFocus()
```
This name is not very descriptive. "Check" is one of the words you should almost never use, because it's so ambiguous.

```suggestion
    void OnFocusLoss()
```

Chase-Percy replied:
Yea, it was pretty bad in hindsight, I did sit there thinking what to name it but yea didn't really work lol. A good reminder of ambiguity though.

Broxzier replied:
Generally it's best to pick a name that describes what a function does, not how it is used. That makes the intent much clearer at first glance.

Broxzier commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -117,17 +128,20 @@ class ServerListWindow final : public Window
     {
         _serverList = {};
         _fetchFuture = {};
+        CheckTextBoxFocus();
```
You could just call `ConfigSaveDefault();` here instead.

Chase-Percy replied:
Feels silly to not have seen that, I was too focused on the concept of losing focus to see the clearer solution: [b147c0d](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/b147c0dfb23ff4a1af469355ba37d0e69c8fdc8c)

ZehMatt commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -267,22 +281,17 @@ class ServerListWindow final : public Window
             return;
 
-        std::string temp = static_cast<std::string>(text);
+        auto temp = static_cast<u8string>(text);
```
This looks a bit dubious, text is `std::string_view`, it should be probably `auto temp = u8string{text};`

Chase-Percy replied:
Yea, good call. It's been a learning experience this PR with not just updating things but evaluating what they are and the better way to implement them: [c9b3e8f](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/c9b3e8f185d96dacacb721875134e639864f3ea4)

Broxzier commented on distribution/changelog.txt:
```diff
@@ -37,6 +37,7 @@
 - Fix: [#19955] Mine Train Roller Coaster has incorrect supports on the sloped left small turn (original bug).
 - Fix: [#19987] [Plugin] ‘SetCheatAction’ has wrong ID in plugin API.
 - Fix: [#19991] Crash using cut-away view.
+- Fix: [#19935] Part of old multiplayer name remains when entering a shorter name.
```
This is a regression, so no changelog entry is necessary.

Chase-Percy replied:
[82cbf57](https://github.com/OpenRCT2/OpenRCT2/pull/20004/commits/82cbf57ec61abe1fa39c3c88b40d44a22632766b)

Broxzier commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -90,7 +92,8 @@ class ServerListWindow final : public Window
     {
-        window_server_list_widgets[WIDX_PLAYER_NAME_INPUT].string = _playerName;
+        _playerName = gConfigNetwork.PlayerName;
+        window_server_list_widgets[WIDX_PLAYER_NAME_INPUT].string = const_cast<utf8*>(_playerName.c_str());
```
This doesn't seem very safe. The buffer that `.c_str()` returns is not guaranteed to stay in place when the string changes. When it moves, this widget will point to garbage.

Broxzier commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -127,7 +129,7 @@ class ServerListWindow final : public Window
                 break;
             case WIDX_PLAYER_NAME_INPUT:
-                WindowStartTextbox(*this, widgetIndex, STR_STRING, _playerName, 63);
+                WindowStartTextbox(*this, widgetIndex, STR_STRING, _playerName.c_str(), MaxPlayerNameLength);
```
Interesting that the max length in this call used to be 63, while the max name length was 32 characters.

Broxzier commented on src/openrct2-ui/windows/ServerList.cpp:
```diff
@@ -267,22 +269,17 @@ class ServerListWindow final : public Window
-                }
+                _playerName = temp;
+                gConfigNetwork.PlayerName = _playerName;
+                window_server_list_widgets[WIDX_PLAYER_NAME_INPUT].string = const_cast<utf8*>(_playerName.c_str());
```
Since the buffer is also set here, I suppose that it's safe, but it's still bad design. One way to ensure that the buffer remains valid is to not use u8string's internal buffer directly.

Since the current implementation works, feel free to just leave this for now.

Chase-Percy replied:
I guess it could just use a std array of uint8_t but yea I was just recommended to use the u8string instead of the original c style stuff that was there. What would be the ideal solution? I'm not sure how changing the underlying union in the widget stuff would affect things so I don't know if the solution would be in that area or not or something like I mentioned.
